minetest/minetest
# Send only relevant ItemStack meta fields to Client

The goal of this PR is to avoid massive network traffic and "bookbanning". This PR prevents sending any ItemStack metadata to the client, except the `description`, `color` and `tool_capabilities`. It was originally created for the Clamity server to prevent "bookbanning" and lag, but then I realized that other servers might also profit from this patch so I might as well contribute it to upstream. Bookbanning works by putting items that contain a lot of data into shulkerboxes (in case you don't know, shulkerboxes are basically portable chests, when they are dug, the items inside get stored in the metadata of the itemstack). When several of these boxes, filled with items with lots of data, are dropped into the the inventory of a player, the inventory gets so large it is too big to send and the server crashes with this message:
```
2021-03-01 04:58:54: ERROR[ConnectionSend]: /home/clamity/minetest-5.4.0/src/network/connection.cpp:3f9: bool con::UDPPeer::processReliableSendCommand(con::ConnectionCommand&, unsigned int): An engine assumption 'c.data.getSize() < 0x8000*512' failed.
```
Every time the affected player tries logging in the server will crash again, with this message.
A similar exploit exists in Minecraft, except that it does not crash the server, but rather kicks the client so it is effectively banned. In Minecraft books are used to ban players using this method; in MineClone2 books have a limit of 5kB, so even if they were filled with random characters (to avoid zlib compression), they would contain less data then e.g. an enchanted pickaxe (21kB).

This PR does not significantly reduce the network traffic that comes from having enchanted tools themselves in the inventory, since the data of enchanted tools is the tool_capabilities, but for shulkerboxes with enchanted stuff, it reduces the data **from 0.5 MB to a few byte**. This does not only help against bookbanning, but also removes lag.

This affects player inventories, detached inventories and inventories in node metadata.

## To do

This PR is a Work in Progress.

- [x] Remove the `name` field from the whitelist, since it is not needed by the client
- [x] Add a possibility to disable this patch globally (to make local map saving possible)
- [x] Add short_description and palette_index to the whitelist 
- [x] Add a possibility to disable this patch for certain players
- [x] Find a solution to fix the problem that the client will sometimes incorrectly predict whether 2 stacks can combine
- [ ] Possibility to mark fields as public

## How to test

Add a clientmod like this:
```lua
minetest.register_chatcommand("wielded", {
	func = function()
		return true, minetest.localplayer:get_wielded_item():to_string()
	end
}) 
```
and a server-side mod like this:
```lua
minetest.register_chatcommand("wielded", {
	func = function(name)
		return true, minetest.get_player_by_name(name):get_wielded_item():to_string()
	end
})
```
Next, you need to obtain an item with metadata. This can be done using the MineClone2 subgame, by putting a few items into a shulkerbox. Then wield the item, and run both the .wielded and /wielded command. You'll notice that the client command will print significantly less data. You can also try it with a named shulkerbox (name it using an anvil) - the colored description, will still work and also show up in the .wielded command, but the serialized inventory data of the shulkerbox does not.
